core/src/main/java/org/infinispan/interceptors/distribution/L1LastChanceInterceptor.java

```diff
@@ -76,40 +76,50 @@ public void start() {
    }
 
    public CompletableFuture<Void> visitDataWriteCommand(InvocationContext ctx, DataWriteCommand command, boolean assumeOriginKeptEntryInL1) throws Throwable {
-      Object returnValue = ctx.forkInvocationSync(command);
-      Object key;
-      if (shouldUpdateOnWriteCommand(command) && command.isSuccessful() &&
-            cdl.localNodeIsOwner((key = command.getKey()))) {
-         if (trace) {
-            log.trace("Sending additional invalidation for requestors if necessary.");
+      ctx.onReturn((ctx1, command1, rv, throwable) -> {
+         if (throwable != null)
+            throw throwable;
+
+         Object key;
+         if (shouldUpdateOnWriteCommand(command) && command.isSuccessful() &&
```

Comment:
TBH I'm not 100% convinced we should keep them either... they should always be the same, they're only here to allow the reuse of `ReturnHandler` instances (e.g. `InvocationContextInterceptor.defaultReturnHandler`). 
This helps reduce the allocation rate, but as you've noticed it also makes `ReturnHandler` lambdas quite a bit uglier.